# Step one: Cleaning

In this notebook we will focus on taking the original data file and reducing all the errors in it. This will produce a clean file that will allow for meaningful analysis.

## Importing libraries and reading the file

In [95]:
import pandas as pd

attacks = pd.read_csv("data/attacks.csv", encoding='unicode_escape')

#We will alos run the 'cleaning.py' script as it will define functions we will use

%run -i 'src/cleaning.py'

## Initial exploration
Now we will start exploring the file. Using the ".head()" we can get a preview of the columns and a the first rows of the dataset.

In [2]:
attacks.head()

Case Number         Date    Year        Type    Country             Area  \
0  2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1  2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2  2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3  2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4  2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   

                         Location     Activity             Name Sex   ...  \
0     Oceanside, San Diego County     Paddling      Julie Wolfe    F  ...   
1  St. Simon Island, Glynn County     Standing  Adyson McNeely     F  ...   
2                    Habush, Oahu      Surfing      John Denges    M  ...   
3              Arrawarra Headland      Surfing             male    M  ...   
4                        La Ticla  Free diving   Gustavo Ramos     M  ...   

          Species           Investigator or Source                       pdf  \
0      White shark                R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1              NaN  K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2              NaN  K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3        2 m shark                  B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4  Tiger shark, 3m                       A .Kipper      2018.06.04-Ramos.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.18   
2  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.09   
3  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.08   
4  http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.04   

  Case Number.2 original order Unnamed: 22 Unnamed: 23  
0    2018.06.25         6303.0         NaN         NaN  
1    2018.06.18         6302.0         NaN         NaN  
2    2018.06.09         6301.0         NaN         NaN  
3    2018.06.08         6300.0         NaN         NaN  
4    2018.06.04         6299.0         NaN         NaN  

[5 rows x 24 columns]

In [3]:
attacks.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

In [4]:
attacks.shape

(25723, 24)

Using 'head()' , '.columns' and '.shape' we can see that we are dealing with a file of 24 columns (names listed above) and 25723 entries. Lets explore further:

In [5]:
attacks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

From this output we gather a few things:

#### Types and names of the columns: 
* There is only two columns with floats
    * This is odd considering that the column "Age" should also be a number
    * The column "Original order" probably adds little to no value
* Columns 0, 19 and 20 all reference a case number
* Columns 17 and 18 could also be duplicated, since they both reference "href"

####  Non-null values:
* Column 0 "Case number" has significantly more non-null values than the rest of the columns
* Columns 22 and 23 are unnamed and have 2 or less non-null values, these columns add no value to the analysis
* Several columns have a non-null count nearing 6300. But most columns have less than this ammount. This means that we are dealing with a great amount of incomplete rows.


### Deleting Unnamed columns:
We will start cleaning the file by deleting the last 2 columns. For this we will use a function declared in the 'cleaning.py' ran in the beginning of this notebook:

In [6]:
delete_columns(attacks,['Unnamed: 22','Unnamed: 23'])

Deleted columns:  ['Unnamed: 22', 'Unnamed: 23']


In [7]:
attacks.sample(5)

Case Number         Date    Year      Type Country    Area  \
2649   1989.12.19  19-Dec-1989  1989.0  Provoked     USA  Hawaii   
7957            0          NaN     NaN       NaN     NaN     NaN   
6606            0          NaN     NaN       NaN     NaN     NaN   
10227         NaN          NaN     NaN       NaN     NaN     NaN   
22307         NaN          NaN     NaN       NaN     NaN     NaN   

                             Location                            Activity  \
2649   90 miles east of Hilo, Hawai'i  On board 51' fishing vessel One Ki   
7957                              NaN                                 NaN   
6606                              NaN                                 NaN   
10227                             NaN                                 NaN   
22307                             NaN                                 NaN   

                 Name Sex   ... Fatal (Y/N) Time Species   \
2649   George Sohswel    M  ...           N  NaN      NaN   
7957              NaN  NaN  ...         NaN  NaN      NaN   
6606              NaN  NaN  ...         NaN  NaN      NaN   
10227             NaN  NaN  ...         NaN  NaN      NaN   
22307             NaN  NaN  ...         NaN  NaN      NaN   

                            Investigator or Source                     pdf  \
2649   J. Borg, p.78; L. Taylor (1993), pp.108-109  1989.12.19-Sohswel.pdf   
7957                                           NaN                     NaN   
6606                                           NaN                     NaN   
10227                                          NaN                     NaN   
22307                                          NaN                     NaN   

                                            href formula  \
2649   http://sharkattackfile.net/spreadsheets/pdf_di...   
7957                                                 NaN   
6606                                                 NaN   
10227                                                NaN   
22307                                                NaN   

                                                    href Case Number.1  \
2649   http://sharkattackfile.net/spreadsheets/pdf_di...    1989.12.19   
7957                                                 NaN           NaN   
6606                                                 NaN           NaN   
10227                                                NaN           NaN   
22307                                                NaN           NaN   

      Case Number.2 original order  
2649     1989.12.19         3654.0  
7957            NaN            NaN  
6606            NaN            NaN  
10227           NaN            NaN  
22307           NaN            NaN  

[5 rows x 22 columns]

### Dealing with the "Case Number" columns
Our data set contains 3 columns referring to a case number. For this we will use a function declared in the 'cleaning.py' ran in the beginning of this notebook: 

In [94]:
compare_columns(attacks,["Case Number","Case Number.1","Case Number.2"])

Column information:


Columns            Case Number  Case Number.1  Case Number.2
Lengths                  25723          25723          25723
Non - null values         8702           6302           6302

Data type comparison:


Case Number  Case Number.1  Case Number.2
<class 'float'>        17021          19421          19421
<class 'str'>           8702           6302           6302

Value comparisons:


Case Number  Case Number.1  Case Number.2
Case Number          25723          23298          23318
Case Number.1        23298          25723          25703
Case Number.2        23318          25703          25723

The above matrix shows the amount of rows that each column shares with other columns. E.g. how many cells store the same information in both columns
Sample of columns:


Case Number Case Number.1 Case Number.2
21204         NaN           NaN           NaN
13777         NaN           NaN           NaN
23856         NaN           NaN           NaN
25417         NaN           NaN           NaN
15509         NaN           NaN           NaN

We can see that the columns "Case Number.1" and "Case Number.2" hold almost the same information (They return the same values in all the tests, and in the comparison matrix they only have 20 values that are different).

In [99]:


def comparison (df, columns):
    tracker = df[0]
    for i in columns:
        tracker = tracker and df[i]
    return tracker


#samples["test"] = samples.apply(lambda row: row)


In [97]:
attacks["Case Number"]

0        2018.06.25
1        2018.06.18
2        2018.06.09
3        2018.06.08
4        2018.06.04
            ...    
25718           NaN
25719           NaN
25720           NaN
25721           NaN
25722            xx
Name: Case Number, Length: 25723, dtype: object